### PR TITLE
Add YOLOX variants to OTX

### PR DESCRIPTION
**Pull request by jaegukhyun**
### Summary
This PR includes adding YOLOX variants to OTX.
Current existed YOLOX is YOLOX-TINY, the smallest YOLOX variant. 
As YOLOX variants show good performance w.r.t accuracy and time, it would be nice to add these YOLOX variants to OTX

Here's some experiments summary.
<html>
<body>

  | YOLOX-TINY | YOLOX_S | YOLOX_L | YOLOX_X | Mobilenet-ATSS | ResNeXt-ATSS | Deformable-DETR | DINO(batch_size = 2)
-- | -- | -- | -- | -- | -- | -- | -- | --
BCCD | 0.89204 | 0.92738 | 0.91809 | 0.91937 | 0.93432 | 0.93183 | 0.93519 | 0.93547
BCCD-SMALL | 0.88939 | 0.95498 | 0.94772 | 0.95075 | 0.91623 | 0.92484 | 0.93639 | 0.8975
FISH | 0.93533 | 0.89212 | 0.90926 | 0.91508 | 0.95663 | 0.96321 | 0.82701 | 0.95005
FISH-SMALL | 0.92424 | 0.77778 | 0.91667 | 0.92857 | 0.83333 | 0.95833 | 0.89583 | 0.95833
POTHOLE | 0.73111 | 0.71877 | 0.7034 | 0.75255 | 0.7971 | 0.82397 | 0.77126 | 0.81455
POTHOLE-SMALL | 0.27465 | 0.32187 | 0.35799 | 0.41236 | 0.43391 | 0.53674 | 0.25408 | 0.39816
VITENS | 0.86252 | 0.86503 | 0.87652 | 0.87972 | 0.87637 | 0.88967 | 0.88476 | 0.91272
VITENS-SMALL | 0.76822 | 0.85981 | 0.87009 | 0.87258 | 0.8391 | 0.87448 | 0.7958 | 0.896
BDD100K | 0.248 | 0.371 | 0.445 | 0.442 | 0.402 | 0.455(Fixed schedule) | 0.448(Fixed schedule) | 0.472(Fixed schedule)

</body>
</html>

### How to test
Intg tests

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [x] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [x] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [x] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: docs/source/guide/explanation/algorithms/object_detection/object_detection.rst
Comment (goodsong81) [inline docs/source/guide/explanation/algorithms/object_detection/object_detection.rst:1]: Why don't we remove `Custom` & `Gen3` tags from model template IDs for latest models that are not yet exposed to Geti? OK for consistency, but we don't need to follow the old rules. (Yeah, it's not Geti rule, but old OTE team's rule)

File: src/otx/algorithms/detection/configs/detection/cspdarknet_yolox_s/deployment.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+_base_ = ["../../base/deployments/base_detection_dynamic.py"]
+
+ir_config = dict(
+    output_names=["boxes", "labels"],
+)
+
+backend_config = dict(
+    model_inputs=[dict(opt_shapes=dict(input=[-1, 3, 416, 416]))],
```
Comment (goodsong81) [inline src/otx/algorithms/detection/configs/detection/cspdarknet_yolox_s/deployment.py]: Do you know why in YOLOX train in 640x640 but infer in 416x416?
c.c. @eunwoosh

Comment (jaegukhyun) [reply src/otx/algorithms/detection/configs/detection/cspdarknet_yolox_s/deployment.py]: backend_config.model_inputs will be configured from data_pipeline.py. Therefore exported model will inference on 640*640 images. However YOLOX-TINY, the existed one, resize 416 * 416 for test data pipeline, so YOLOX-TINY exported model will inference on 416 * 416 image. I guess this setting is for faster inference since YOLOX-TINY is Geti's speed model.
Anyway @goodsong81 do you think this file should be modified even if it will automatically be patched?

File: docs/source/guide/explanation/algorithms/object_detection/object_detection.rst
Comment (jaegukhyun) [reply docs/source/guide/explanation/algorithms/object_detection/object_detection.rst:1]: I changed template id for YOLOX variants, I'll change remaining models' template id in another PR.

File: src/otx/algorithms/detection/configs/detection/cspdarknet_yolox_s/deployment.py (hunk 1 above)
Comment (eunwoosh) [reply src/otx/algorithms/detection/configs/detection/cspdarknet_yolox_s/deployment.py]: I think the reason why YOLOX tiny uses 416 input size in test pipeline is because config was just copied from mmdet. I don't know why mmdet uses that input size, but it's weird to use 640 input size when training and use 412 at inference. Anyway, I think although it's overwritten by 'data_pipeline.py', but changing it as 640 can be better to avoid possible confusion later.

**Review by goodsong81 — COMMENTED** [review]
I prefer the task type prefix like `Object_Detection_YOLOX_S` but open to discussion. Anyway, we would follow some convention afterwards.
@sungmanc @eunwoosh What do you think?

**Comment by jaegukhyun** [comment]
> I prefer the task type prefix like `Object_Detection_YOLOX_S` but open to discussion. Anyway, we would follow some convention afterwards. @sungmanc @eunwoosh What do you think?

I agree on having prefix for unity. Object_Detection seems fine.

**Comment by sungmanc** [comment]
> I prefer the task type prefix like `Object_Detection_YOLOX_S` but open to discussion. Anyway, we would follow some convention afterwards. @sungmanc @eunwoosh What do you think?

I also prefer to use the task type prefix. Sometimes, it could be helpful to identify easily.

**Review by sungmanc — DISMISSED** [review]

**Review by goodsong81 — DISMISSED** [review]

File: src/otx/algorithms/detection/configs/detection/cspdarknet_yolox_l/hpo_config.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+metric: mAP
+search_algorithm: asha
+early_stop: None
+hp_space:
+  learning_parameters.learning_rate:
+    param_type: qloguniform
+    range:
+      - 0.00002
+      - 0.002
+      - 0.00001
```
Comment (eunwoosh) [inline src/otx/algorithms/detection/configs/detection/cspdarknet_yolox_l/hpo_config.yaml]:
```suggestion
      - 0.0001
      - 0.01
      - 0.0001
```
I suggest you to change values as above.

File: src/otx/algorithms/detection/configs/detection/cspdarknet_yolox_s/hpo_config.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+metric: mAP
+search_algorithm: asha
+early_stop: None
+hp_space:
+  learning_parameters.learning_rate:
+    param_type: qloguniform
+    range:
+      - 0.00002
+      - 0.002
+      - 0.00001
```
Comment (eunwoosh) [inline src/otx/algorithms/detection/configs/detection/cspdarknet_yolox_s/hpo_config.yaml]:
```suggestion
      - 0.0001
      - 0.01
      - 0.0001
```
same as above

File: src/otx/algorithms/detection/configs/detection/cspdarknet_yolox_x/hpo_config.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+metric: mAP
+search_algorithm: asha
+early_stop: None
+hp_space:
+  learning_parameters.learning_rate:
+    param_type: qloguniform
+    range:
+      - 0.00002
+      - 0.002
+      - 0.00001
```
Comment (eunwoosh) [inline src/otx/algorithms/detection/configs/detection/cspdarknet_yolox_x/hpo_config.yaml]:
```suggestion
      - 0.0001
      - 0.01
      - 0.0001
```
same as above

**Review by eunwoosh — COMMENTED** [review]
LGTM but I think HPO configuration need to be changed. Please take a look.

**Comment by eunwoosh** [comment]
> I prefer the task type prefix like `Object_Detection_YOLOX_S` but open to discussion. Anyway, we would follow some convention afterwards. @sungmanc @eunwoosh What do you think?

Yes I agree to use the prefix they said.

**Review by eunwoosh — DISMISSED** [review]

**Review by sungmanc — APPROVED** [review]

**Review by goodsong81 — APPROVED** [review]
